### PR TITLE
Added reversed for proof-of-stake syncing

### PR DESCRIPTION
**Pull request by Giulio2002**

### CODE REVIEW

**Comment by Giulio2002** [comment]
Note: this is just an experimental version, it does not make use of request skeleton and it can roughly download 4000 blocks/second... treat this as being a mere prototype, it was able to sync all the way down from block 13,665,338 in 57 minutes roughly, all subsequent stages are executed correctly.

**Comment by AlexeyAkhunov** [comment]
Sorry, I introduced some conflicts for your with my simplification PR. Once you resolved, I will review this one again

**Comment by Giulio2002** [comment]
If anyone wants to test it 0, you can set terminalTotalDifficulty in Mainnet chainconfig and run erigon on mainned with exposed privateapi, after that run rpcdaemon with `--http.api=engine`and in another terminal use the following command : `curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"engine_executePayloadV1","params":[{"parentHash":"0x49996e2ca45c30dd2c4692afe2053b6ece6b22c78b2d95877144708e64d1eb13","coinbase":"0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b","stateRoot":"0xca3149fa9e37db08d1cd49c9061db1002ef1cd58db2210f2115c8c989b2bdf45","receiptRoot":"0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421","logsBloom":"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","random":"0x0000000000000000000000000000000000000000000000000000000000000000","blockNumber":"0xd1efb7","gasLimit":"0x1c9c380","gasUsed":"0x0","timestamp":"0x5","extraData":"0x","baseFeePerGas":"0x7","blockHash":"0xf9218048a47240d048dff4a22513cda5968b3c460d82e873273139b8973426f4","transactions":[]}],"id":67}' http://localhost:8545

`

**Comment by Giulio2002** [comment]
there is actually a way to make it way faster (aka 20k blk/sec) but for some reason i have a bug in my other implementation which consist in some request not going through even though the request object is formed, note that updating the next request is done sentPeers is true but for some reason after a while my other implementation get stuck and does not receive some left out blocks

**Comment by Giulio2002** [comment]
There is also the advanced unwind logic, if we did not find a parenthash in a database past our previous progress, we unwind due to the fact that we are likely to have forked.

File: cmd/sentry/download/downloader.go
Comment (AlexeyAkhunov) [inline cmd/sentry/download/downloader.go:474]: Can you restructure this as:
```
if penaltyKind == header download.NoPenatly {
   if cs.Hd.GetBackwards() {
      ....
   } else {
       ....
   }
else {
 ...
}
```

File: turbo/stages/headerdownload/header_algos.go
Comment (AlexeyAkhunov) [inline turbo/stages/headerdownload/header_algos.go:578]: Is `anchorQueue` relevant here?

File: turbo/stages/headerdownload/header_algos.go
Comment (AlexeyAkhunov) [inline turbo/stages/headerdownload/header_algos.go:576]: If this method does not change anything inside `hd`, you should do `hd.lock.RLock()` and `hd.lock.RUnlock` instead

File: turbo/stages/headerdownload/header_algos.go
Comment (AlexeyAkhunov) [inline turbo/stages/headerdownload/header_algos.go:603]: If this function is every called for the `HeaderRequest` which is created by the function `RequestMoreHeadersForPOS`, the `req.Anchor.timeouts++` will panic, because those requests do not have anchor associated with them.

File: eth/stagedsync/stage_headers.go
Comment (AlexeyAkhunov) [inline eth/stagedsync/stage_headers.go:240]: You need to make sure these collectors are closed (and files wiped out as a result) if this ends up with error. I suggest adding `defer`s

File: eth/stagedsync/stage_headers.go
Comment (AlexeyAkhunov) [inline eth/stagedsync/stage_headers.go:248]: I suggest not reusing this function here, since it deals with anchor you don't have. Create a separate function for what you need to do

File: eth/stagedsync/stage_headers.go
Comment (AlexeyAkhunov) [inline eth/stagedsync/stage_headers.go:253]: Look like you are just repeating the code 2 times. Why not make a loop and limit number of repetitions perhaps?

File: turbo/stages/headerdownload/header_algos.go
Comment (AlexeyAkhunov) [inline turbo/stages/headerdownload/header_algos.go:575]: This function does not need `currentTime`

File: turbo/stages/headerdownload/header_algos.go
Comment (AlexeyAkhunov) [inline turbo/stages/headerdownload/header_algos.go:786]: you can bypass `insertList` and just do insertion into the collectors right here. Basically, move the content of `InsertHeadersBackwards` into this one. Since headers don't go directly into DB, but into collectors, it should be pretty quick, and you don't need to allocate more objects for links

File: turbo/stages/headerdownload/header_algos.go
Comment (Giulio2002) [reply turbo/stages/headerdownload/header_algos.go:603]: yes, sorry i did not try it with the simplifications you put in

File: turbo/stages/headerdownload/header_algos.go
Comment (Giulio2002) [reply turbo/stages/headerdownload/header_algos.go:786]: i also want to check if the headers match with their parenthashes to check their validity, so i need to process them in order, not all requests are received in the order we sent them, i still will need to keep them somewhere.

Comment (Giulio2002) [reply turbo/stages/headerdownload/header_algos.go:787]: that is why, i first wait for the previous request to go in before moving on

File: turbo/stages/headerdownload/header_algos.go
Comment (AlexeyAkhunov) [inline turbo/stages/headerdownload/header_algos.go:731]: It is probably better to just iterate over the `insertList` and make it empty after the loop than to do re-slicing

Comment (AlexeyAkhunov) [reply turbo/stages/headerdownload/header_algos.go:731]: I mean this:
```
for _, header := range hd.insertList {
...
}
hd.insertList = hd.insertList[:0]
```
it is more efficient and more readable that way

File: turbo/stages/headerdownload/header_algos.go
Comment (AlexeyAkhunov) [inline turbo/stages/headerdownload/header_algos.go:742]: If the hash does not match, you don't really want to keep anything in the `insertList`, so approach I suggested above would work

File: turbo/stages/headerdownload/header_algos.go
Comment (AlexeyAkhunov) [reply turbo/stages/headerdownload/header_algos.go:787]: Ok, understood. One small improvement you can make is to add `headerRaw` into the link you are creating, so you don't have to do RLP encoding again in the `InsertHeadersBackwards`. I introduced this after you've first created this PR

File: eth/stagedsync/stage_headers.go
Comment (AlexeyAkhunov) [inline eth/stagedsync/stage_headers.go:229]: `stopped` never set to `true` inside this loop. Also, this looks like a "tight loop", meaning that it will churn out requests and try to push them to the sentry, without any pauses, this will eat CPU and also put a lot of pressure on the communication with the sentry. You need some kind of back off logic for the case when `sentToPeer` is false, to let peers respond to previous requests. Look how it is done for forward sync, I believe there is 1 second timer

File: eth/stagedsync/stage_headers.go
Comment (AlexeyAkhunov) [inline eth/stagedsync/stage_headers.go:231]: Sorry, it is this loop that will be mostly tight

File: turbo/stages/headerdownload/header_algos.go
Comment (AlexeyAkhunov) [inline turbo/stages/headerdownload/header_algos.go:725]: You can just use `segmentFragment.HeaderRaw` instead of re-encoding

File: eth/stagedsync/stage_headers.go
Comment (AlexeyAkhunov) [inline eth/stagedsync/stage_headers.go:226]: You can expand this defer to also unset collectors from `hd`, to allow them to be garbage-collected after this stage is complete, like so:
```
defer func() {
   cfg.hd.SetHeadersCollector(nil)
   cfg.hd.SetCanonicalHashesCollector(nil)
}()
```

File: eth/stagedsync/stage_headers.go
Comment (AlexeyAkhunov) [reply eth/stagedsync/stage_headers.go:229]: You can also check out in the forward sync how `ctx.Done()` is used to set `stopped = false` so that this process can be gracefully broken by Ctrl-C or by sending `SIGTERM` signal (which will lead to closing of the context `ctx`). And such check also need to happen within sub loop below

File: eth/stagedsync/stage_headers.go
Comment (AlexeyAkhunov) [reply eth/stagedsync/stage_headers.go:231]: That is one of the reason in the forward sync I limited number of iterations to 64, and also inserted `select` clause with timer, with `ctx.Done()` and also with printing progress. I recommend you do it also here

Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go:231]: if i used a 30 seconds timer i would get 6.8 blk/sec, what i could do is `sentToPeer==false ?` then i sleep for 30 milliseconds, by doing it i dont implode the CPU and requests can be handled just fine. also i set stopped to true in a goroutine like it is done in other stages.

File: eth/stagedsync/stage_headers.go
Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go:236]: i do it in the goroutine, now i also check it in the request loop

Comment (AlexeyAkhunov) [reply eth/stagedsync/stage_headers.go:236]: It is usually not advised to use a variable from more than one goroutine without any synchronisation. Even if the variable can be atomically changed. Because of any possible caching. You need to either use mutexes, or atomics, or channels, or remove that extra goroutine.

File: eth/stagedsync/stage_headers.go
Comment (AlexeyAkhunov) [inline eth/stagedsync/stage_headers.go:242]: this is ok, but I would consider it low-tech :)

File: eth/stagedsync/stage_headers.go
Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go:235]: now i am using channels

File: eth/stagedsync/stage_headers.go
Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go]: I am now using a timer like in headers forward, is this more high-tech?

File: eth/stagedsync/stage_headers.go
Comment (AlexeyAkhunov) [reply eth/stagedsync/stage_headers.go:235]: Ok, that should work

**Review by AlexeyAkhunov — APPROVED** [review]
That's probably enough commenting for now, I think we can start with that. But there is still room for improvements :)

File: eth/stagedsync/stage_headers.go
Comment (AlexeyAkhunov) [reply eth/stagedsync/stage_headers.go]: ok :) I think we can stop at that. What I meant is combining the timer with other things, like `ctx.Done()` so that you don't need to run a extra goroutine, but we can change it in a separate PR

Comment (AlexeyAkhunov) [reply eth/stagedsync/stage_headers.go]: The main reason I said it was low-tech, is because `time.Sleep` cannot be interrupted and composed with other quitting conditions. It would be better to use one `select` that handles various way this can be interrupted (on timer, on context closure). But as I said we can do it separately as refactoring

File: cmd/sentry/download/downloader.go
Comment (yperbasis) [inline cmd/sentry/download/downloader.go:476]: I'm a bit confused about this backwards flag. Firstly, my understanding is that the existing (pre-Merge) downloader downloads headers mostly (besides skeleton requests) backwards, towards the genesis. So "backwards" meaning PoS sync is a bit misleading. Secondly, why is it all or nothing mode? Couldn't exclusively backwards downloading of PoS blocks run simultaneously with the old skeleton+backwards algorithm for pre-Merge blocks? To sum up, to my mind "GetBackwards" is a confusing name.

Comment (Giulio2002) [reply cmd/sentry/download/downloader.go]: I changed it to POSSync, the flag indicates when we are doing sync for proof-of-stake or not

Comment (yperbasis) [reply cmd/sentry/download/downloader.go]: So does it mean that we can't download PoS and PoW headers simultaneously?

Comment (Giulio2002) [reply cmd/sentry/download/downloader.go]: we can, if we assume that the block supplied by the beacon chain is correct, we sync backwards until we find a block whose parenthash is present in our database, this can be past the terminal block. if such thing happen, we unwind

File: eth/stagedsync/stage_headers.go
Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go]: oh that is what you meant, okay, now i understand, now it does not have the goroutine anymore

File: cmd/sentry/download/downloader.go
Comment (yperbasis) [reply cmd/sentry/download/downloader.go]: Hmm. What about a node that wants to sync from scratch some time after the Merge? It doesn't have any PoW headers in the DB yet.

Comment (Giulio2002) [reply cmd/sentry/download/downloader.go]: It will first sync forward like we do already normally in POW, and after we surpass totalTerminalDifficulty, the node will stand still and wait for payloads and then proceed to backward sync once it receives a payload. a while after the merge we may even remove forward sync and have just downward sync.

Comment (Giulio2002) [reply cmd/sentry/download/downloader.go]: You can try this, if you follow the procedure i pointed out above in the PR.

**Review by yperbasis — CHANGES_REQUESTED** [review]
There is some git merge/rebase issue in the PR: I see unrelated stuff like `ParliaConfig`

File: cmd/sentry/download/downloader.go
Comment (yperbasis) [reply cmd/sentry/download/downloader.go]: Well, the problem is that our current (PoW) downloader is not a forward one (and the name `HeadersForward` is incorrect). As far as I understand it, the current downloader first sends a skeleton request and then starts extending resulting anchors downwards, towards genesis. So it's not correct that we first download in the forward direction, then reach the Terminal Total Difficulty (TTD), and then switch to the PoS header downloader in the downwards direction. And after the merge the skeleton request will likely return some PoS blocks pass the TTD, won't it?

Comment (Giulio2002) [reply cmd/sentry/download/downloader.go]: Okay, i think that headersForward means that headers are inserted in "forward" order, if you were remove preverified hashes, you would have to consensus check in forward order(genesis to head), however in our case we process the headers in downward order we start from head and go to genesis

Comment (Giulio2002) [reply cmd/sentry/download/downloader.go]: If you think that `headersFoward` is wrong/misleading that can be refactored in another PR, it is not in the scope of this one i believe

Comment (Giulio2002) [reply cmd/sentry/download/downloader.go]: We need to insert them backwards to check their validity, we need to check if previous block parenthash is equal to current block hash

**Pull request merged**
